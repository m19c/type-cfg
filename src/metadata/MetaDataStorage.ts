import { ensureReflectMetadataExists } from './util';
import { DefinitionMetadata } from './definitions/DefinitionMetadata';
import { PropertyMetadata } from './definitions/PropertyMetadata';
import { EnumMetadata } from './definitions/EnumMetadata';

export class MetadataStorage {
  private definitions: DefinitionMetadata[] = [];
  private properties: PropertyMetadata[] = [];
  private enums: EnumMetadata[] = [];

  constructor() {
    ensureReflectMetadataExists();
  }

  collectDefinitionMetadata(item: DefinitionMetadata) {
    this.definitions.push(item);
  }

  findDefinitionByInstance(value: any) {
    return (
      this.definitions.filter(definition => value instanceof definition.target).shift() || null
    );
  }

  findDefinitionByValue(value: any) {
    return this.definitions.filter(definition => value === definition.target).shift() || null;
  }

  collectConfigPropertyMetadata(item: PropertyMetadata) {
    this.properties.push(item);
  }

  findPropertiesByInstance(value: any) {
    return this.properties.filter(property => value instanceof property.target);
  }

  collectEnumMetadata(item: EnumMetadata) {
    this.enums.push(item);
  }

  findEnum(value: any) {
    return this.enums.filter(item => item.value === value).shift() || null;
  }
}
