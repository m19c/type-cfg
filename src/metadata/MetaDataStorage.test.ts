/* eslint @typescript-eslint/no-non-null-assertion: 0 */

import { MetadataStorage } from './MetaDataStorage';

jest.mock('./util.ts');
// eslint-disable-next-line @typescript-eslint/no-var-requires
const util = require('./util');

describe('metadata/MetaDataStorage', () => {
  test('constructor', () => {
    const mds = new MetadataStorage();
    expect(mds).toBeInstanceOf(MetadataStorage);

    expect(util.ensureReflectMetadataExists).toHaveBeenCalledTimes(1);
    util.ensureReflectMetadataExists.mockRestore();
  });

  describe('collect', () => {
    let mds: null | MetadataStorage = null;

    beforeEach(() => {
      mds = new MetadataStorage();
    });

    afterEach(() => {
      mds = null;
    });

    test('collectDefinitionMetadata', () => {
      mds!.collectDefinitionMetadata({
        name: 'Test',
        target: () => {},
      });

      expect(mds).toMatchSnapshot();
    });

    test('collectPropertyMetadata', () => {
      mds!.collectPropertyMetadata({
        name: 'Test',
        target: () => {},
        source: 'TEST_ENV_VAR',
        defaultValue: null,
        delimiter: null,
        determineType: () => {},
        determineTypeOptions: { isArray: true },
        required: false,
      });

      expect(mds).toMatchSnapshot();
    });

    test('collectEnumMetadata', () => {
      mds!.collectEnumMetadata({
        name: 'Test',
        value: {
          A: 1,
          B: 2,
        },
      });

      expect(mds).toMatchSnapshot();
    });
  });

  describe('find', () => {
    class Config {}
    enum Environment {
      Development = 'development',
      Production = 'production',
    }

    let mds: null | MetadataStorage;
    beforeAll(() => {
      mds = new MetadataStorage();

      mds!.collectDefinitionMetadata({
        name: 'Config',
        target: Config,
      });

      mds!.collectPropertyMetadata({
        name: 'environment',
        target: Config,
        source: 'NODE_ENV',
        defaultValue: null,
        delimiter: null,
        determineType: () => Environment,
        determineTypeOptions: { isArray: false },
        required: false,
      });

      mds!.collectPropertyMetadata({
        name: 'domain',
        target: Config,
        source: 'DOMAIN',
        defaultValue: null,
        delimiter: null,
        determineType: () => String,
        determineTypeOptions: { isArray: false },
        required: false,
      });

      mds!.collectEnumMetadata({
        name: 'Environment',
        value: Environment,
      });
    });

    afterAll(() => {
      mds = null;
    });

    test('findEnum', () => {
      expect(mds!.findEnum('notFound')).toBeNull();
      expect(mds!.findEnum(Environment)).not.toBeNull();
    });

    test('findDefinitionByInstance', () => {
      class NotFound {}
      const nfi = new NotFound();

      expect(mds!.findDefinitionByInstance(nfi)).toBeNull();
      expect(mds!.findDefinitionByInstance(new Config())).not.toBeNull();
    });

    test('findDefinitionByValue', () => {
      class NotFound {}

      expect(mds!.findDefinitionByValue(NotFound)).toBeNull();
      expect(mds!.findDefinitionByValue(Config)).not.toBeNull();
    });

    test('findPropertiesByInstance', () => {
      class NotFound {}

      expect(mds!.findPropertiesByInstance(new NotFound())).toHaveLength(0);
      expect(mds!.findPropertiesByInstance(new Config())).toHaveLength(2);
    });
  });
});
