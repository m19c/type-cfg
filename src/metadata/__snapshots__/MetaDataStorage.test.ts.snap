// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`metadata/MetaDataStorage collect collectDefinitionMetadata 1`] = `
MetadataStorage {
  "definitions": Array [
    Object {
      "name": "Test",
      "target": [Function],
    },
  ],
  "enums": Array [],
  "properties": Array [],
}
`;

exports[`metadata/MetaDataStorage collect collectEnumMetadata 1`] = `
MetadataStorage {
  "definitions": Array [],
  "enums": Array [
    Object {
      "name": "Test",
      "value": Object {
        "A": 1,
        "B": 2,
      },
    },
  ],
  "properties": Array [],
}
`;

exports[`metadata/MetaDataStorage collect collectPropertyMetadata 1`] = `
MetadataStorage {
  "definitions": Array [],
  "enums": Array [],
  "properties": Array [
    Object {
      "defaultValue": null,
      "delimiter": null,
      "determineType": [Function],
      "determineTypeOptions": Object {
        "isArray": true,
      },
      "name": "Test",
      "required": false,
      "source": "TEST_ENV_VAR",
      "target": [Function],
    },
  ],
}
`;
